File: agent/src/utils/health.ts
import { logger } from './logger.js';

export interface HealthStatus {
  name: string;
  healthy: boolean;
  message: string;
  details?: Record<string, unknown>;
  timestamp: string;
}

export interface SystemHealth {
  overallHealthy: boolean;
  checks: HealthStatus[];
  uptimeSeconds: number;
  version: string;
  nodeVersion: string;
  platform: string;
}

export class HealthChecker {
  private startTime: number;
  private checks: Map<string, () => HealthStatus> = new Map();
  private version: string;

  constructor(version: string = '1.0.0') {
    this.startTime = Date.now();
    this.version = version;
    this.registerDefaultChecks();
  }

  private registerDefaultChecks(): void {
    this.registerCheck('memory', this.checkMemory.bind(this));
    this.registerCheck('event_loop', this.checkEventLoop.bind(this));
    this.registerCheck('python_bridge', this.checkPythonBridge.bind(this));
  }

  registerCheck(name: string, checkFn: () => HealthStatus): void {
    this.checks.set(name, checkFn);
  }

  runCheck(name: string): HealthStatus {
    const checkFn = this.checks.get(name);
    if (!checkFn) {
      return {
        name,
        healthy: false,
        message: `Unknown check: ${name}`,
        timestamp: new Date().toISOString(),
      };
    }

    try {
      return checkFn();
    } catch (error) {
      return {
        name,
        healthy: false,
        message: `Check failed: ${error}`,
        timestamp: new Date().toISOString(),
      };
    }
  }

  runAllChecks(): SystemHealth {
    const results: HealthStatus[] = [];

    for (const name of this.checks.keys()) {
      results.push(this.runCheck(name));
    }

    const allHealthy = results.every(r => r.healthy);

    return {
      overallHealthy: allHealthy,
      checks: results,
      uptimeSeconds: (Date.now() - this.startTime) / 1000,
      version: this.version,
      nodeVersion: process.version,
      platform: process.platform,
    };
  }

  runQuickCheck(): boolean {
    const criticalChecks = ['memory', 'event_loop'];
    return criticalChecks.every(name => this.runCheck(name).healthy);
  }

  private checkMemory(): HealthStatus {
    const memUsage = process.memoryUsage();
    const heapUsedMB = memUsage.heapUsed / 1024 / 1024;
    const heapTotalMB = memUsage.heapTotal / 1024 / 1024;
    const rssMB = memUsage.rss / 1024 / 1024;
    const percentUsed = (heapUsedMB / heapTotalMB) * 100;

    const healthy = percentUsed < 90;
    let message = `Heap: ${heapUsedMB.toFixed(1)}MB / ${heapTotalMB.toFixed(1)}MB (${percentUsed.toFixed(1)}%)`;

    if (percentUsed > 80) {
      message += ' [WARNING: High memory usage]';
    }

    return {
      name: 'memory',
      healthy,
      message,
      details: {
        heapUsedMB: Math.round(heapUsedMB * 100) / 100,
        heapTotalMB: Math.round(heapTotalMB * 100) / 100,
        rssMB: Math.round(rssMB * 100) / 100,
        percentUsed: Math.round(percentUsed * 100) / 100,
      },
      timestamp: new Date().toISOString(),
    };
  }

  // --- Event loop lag measurement (async-tracked, sync-reported) ---
  private _lastEventLoopLagMs = -1;
  private _eventLoopCheckCount = 0;
  private _measuringEventLoop = false;

  private _measureEventLoopLag(): void {
    if (this._measuringEventLoop) return;
    this._measuringEventLoop = true;

    const start = process.hrtime.bigint();
    setImmediate(() => {
      const elapsed = Number(process.hrtime.bigint() - start) / 1e6; // ms
      // Exponential moving average to smooth outliers
      if (this._lastEventLoopLagMs < 0) {
        this._lastEventLoopLagMs = elapsed;
      } else {
        this._lastEventLoopLagMs = this._lastEventLoopLagMs * 0.7 + elapsed * 0.3;
      }
      this._measuringEventLoop = false;
    });
  }

  private checkEventLoop(): HealthStatus {
    // Trigger a fresh measurement each time this is called
    this._measureEventLoopLag();
    this._eventLoopCheckCount++;

    // Report the most recent measurement (will be -1 on first call, improving over time)
    const lagMs = this._lastEventLoopLagMs < 0 ? 0 : Math.round(this._lastEventLoopLagMs * 10) / 10;

    // Thresholds: <10ms good, <50ms warning, >50ms degraded
    const healthy = this._lastEventLoopLagMs < 0 || lagMs < 50;
    const isWarning = lagMs >= 10 && lagMs < 50;
    const isDegraded = lagMs >= 50;

    let message: string;
    if (this._lastEventLoopLagMs < 0) {
      message = 'Event loop: calibrating (measurements start on next call)';
    } else if (isDegraded) {
      message = `Event loop lag: ${lagMs.toFixed(1)}ms [DEGRADED]`;
    } else if (isWarning) {
      message = `Event loop lag: ${lagMs.toFixed(1)}ms [WARNING]`;
    } else {
      message = `Event loop lag: ${lagMs.toFixed(1)}ms`;
    }

    return {
      name: 'event_loop',
      healthy,
      message,
      details: { lagMs: this._lastEventLoopLagMs < 0 ? null : lagMs, checkCount: this._eventLoopCheckCount },
      timestamp: new Date().toISOString(),
    };
  }

  // --- Python bridge connectivity check ---
  private _bridgeHealthChecked = false;
  private _bridgeHealthy = false;
  private _bridgeMessage = 'Not yet checked';
  private _bridgeDetails: Record<string, unknown> = {};

  private async _checkPythonBridgeAsync(): Promise<void> {
    const httpMode = (process.env.CORE_BRIDGE_MODE || 'http') !== 'subprocess';
    const coreUrl = process.env.CORE_API_URL || 'http://localhost:8000';

    if (httpMode) {
      try {
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), 3000);
        const resp = await fetch(`${coreUrl}/health`, { signal: controller.signal });
        clearTimeout(timer);

        if (resp.ok) {
          const data = (await resp.json()) as Record<string, unknown>;
          this._bridgeHealthy = true;
          this._bridgeMessage = `Python bridge healthy (HTTP ${coreUrl}, v${data.version ?? '?'})`;
          this._bridgeDetails = { mode: 'http', status: resp.status, ...data };
        } else {
          this._bridgeHealthy = false;
          this._bridgeMessage = `Python bridge returned ${resp.status} (${coreUrl})`;
          this._bridgeDetails = { mode: 'http', status: resp.status };
        }
      } catch (err: unknown) {
        this._bridgeHealthy = false;
        this._bridgeMessage = `Python bridge unreachable (${coreUrl}): ${err instanceof Error ? err.message : String(err)}`;
        this._bridgeDetails = { mode: 'http' };
      }
    } else {
      // Subprocess mode: check if a Python process exists and is responsive
      // We can't easily test the subprocess synchronously, so check if Python is available on PATH
      try {
        const { spawnSync } = await import('child_process');
        const pythonCmd = process.env.PYTHON_COMMAND || 'python3';
        const result = spawnSync(pythonCmd, ['-c', 'import sys; print(sys.version)'], {
          timeout: 3000,
          encoding: 'utf8',
        });

        if (result.status === 0 && result.stdout) {
          this._bridgeHealthy = true;
          this._bridgeMessage = `Python bridge available (${pythonCmd}: ${result.stdout.trim().split('\n')[0]})`;
          this._bridgeDetails = { mode: 'subprocess', pythonVersion: result.stdout.trim() };
        } else {
          this._bridgeHealthy = false;
          this._bridgeMessage = `Python command failed (${pythonCmd}): ${result.stderr?.trim() || result.error?.message || 'unknown error'}`;
          this._bridgeDetails = { mode: 'subprocess', exitCode: result.status };
        }
      } catch (err: unknown) {
        this._bridgeHealthy = false;
        this._bridgeMessage = `Python bridge check failed: ${err instanceof Error ? err.message : String(err)}`;
        this._bridgeDetails = { mode: 'subprocess' };
      }
    }
    this._bridgeHealthChecked = true;
  }

  private checkPythonBridge(): HealthStatus {
    // Trigger async check (results appear on next call)
    void this._checkPythonBridgeAsync();

    const details = this._bridgeDetails;
    return {
      name: 'python_bridge',
      healthy: this._bridgeHealthChecked ? this._bridgeHealthy : true, // First call: assume healthy until verified
      message: this._bridgeHealthChecked ? this._bridgeMessage : 'Python bridge: checking (async)...',
      details: {
        ...details,
        checked: this._bridgeHealthChecked,
      },
      timestamp: new Date().toISOString(),
    };
  }

  isHealthy(): boolean {
    return this.runQuickCheck();
  }

  getUptime(): number {
    return (Date.now() - this.startTime) / 1000;
  }
}

export class LivenessProbe {
  private lastHeartbeat: number;
  private timeoutMs: number;

  constructor(timeoutMs: number = 30000) {
    this.timeoutMs = timeoutMs;
    this.lastHeartbeat = Date.now();
  }

  heartbeat(): void {
    this.lastHeartbeat = Date.now();
  }

  isAlive(): boolean {
    return Date.now() - this.lastHeartbeat < this.timeoutMs;
  }

  check(): { alive: boolean; lastHeartbeat: string; secondsSinceHeartbeat: number } {
    return {
      alive: this.isAlive(),
      lastHeartbeat: new Date(this.lastHeartbeat).toISOString(),
      secondsSinceHeartbeat: (Date.now() - this.lastHeartbeat) / 1000,
    };
  }
}

export class ReadinessProbe {
  private ready: boolean = true;
  private reasons: string[] = [];

  setReady(ready: boolean, reason?: string): void {
    this.ready = ready;
    if (reason && !ready) {
      this.reasons.push(reason);
    } else if (ready) {
      this.reasons = [];
    }
  }

  isReady(): boolean {
    return this.ready;
  }

  check(): { ready: boolean; reasons: string[] } {
    return {
      ready: this.ready,
      reasons: this.reasons,
    };
  }
}

export const healthChecker = new HealthChecker();
export const livenessProbe = new LivenessProbe();
export const readinessProbe = new ReadinessProbe();

File: core/src/scholardevclaw/utils/health.py
from __future__ import annotations

import os
import sys
import time
import traceback
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any, Callable

import psutil


@dataclass
class HealthStatus:
    name: str
    healthy: bool
    message: str
    details: dict[str, Any] = field(default_factory=dict)
    timestamp: str = field(default_factory=lambda: datetime.now().isoformat())


@dataclass
class SystemHealth:
    overall_healthy: bool
    checks: list[HealthStatus]
    uptime_seconds: float
    version: str
    python_version: str
    platform: str


class HealthChecker:
    """Comprehensive health check system for ScholarDevClaw."""

    def __init__(self, start_time: float | None = None):
        self.start_time = start_time or time.time()
        self._checks: dict[str, Callable[[], HealthStatus]] = {}
        self._register_default_checks()

    def _register_default_checks(self) -> None:
        self.register_check("python_version", self._check_python_version)
        self.register_check("memory", self._check_memory)
        self.register_check("disk", self._check_disk)
        self.register_check("environment", self._check_environment)
        self.register_check("filesystem", self._check_filesystem)

    def register_check(self, name: str, check_func: Callable[[], HealthStatus]) -> None:
        self._checks[name] = check_func

    def run_check(self, name: str) -> HealthStatus:
        if name not in self._checks:
            return HealthStatus(
                name=name,
                healthy=False,
                message=f"Unknown check: {name}",
            )

        try:
            return self._checks[name]()
        except Exception as e:
            return HealthStatus(
                name=name,
                healthy=False,
                message=f"Check failed: {e}",
                details={"error": str(e), "traceback": traceback.format_exc()},
            )

    def run_all_checks(self) -> SystemHealth:
        results = [self.run_check(name) for name in self._checks]
        all_healthy = all(r.healthy for r in results)

        return SystemHealth(
            overall_healthy=all_healthy,
            checks=results,
            uptime_seconds=time.time() - self.start_time,
            version=self._get_version(),
            python_version=f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
            platform=sys.platform,
        )

    def run_quick_check(self) -> bool:
        critical_checks = ["memory", "disk", "filesystem"]
        return all(self.run_check(name).healthy for name in critical_checks)

    def _get_version(self) -> str:
        try:
            from scholardevclaw import __version__

            return __version__
        except Exception:
            return "unknown"

    def _check_python_version(self) -> HealthStatus:
        version = sys.version_info
        if version >= (3, 10):
            return HealthStatus(
                name="python_version",
                healthy=True,
                message=f"Python {version.major}.{version.minor}.{version.micro}",
            )
        return HealthStatus(
            name="python_version",
            healthy=False,
            message=f"Python {version.major}.{version.minor} not supported. Need 3.10+",
        )

    def _check_memory(self) -> HealthStatus:
        memory = psutil.virtual_memory()
        available_gb = memory.available / (1024**3)
        total_gb = memory.total / (1024**3)
        percent_used = memory.percent

        healthy = percent_used < 90
        message = f"{available_gb:.1f}GB available of {total_gb:.1f}GB ({percent_used}% used)"

        if percent_used > 80:
            message += " [WARNING: High memory usage]"

        return HealthStatus(
            name="memory",
            healthy=healthy,
            message=message,
            details={
                "available_gb": round(available_gb, 2),
                "total_gb": round(total_gb, 2),
                "percent_used": percent_used,
            },
        )

    def _check_disk(self, path: str = "/") -> HealthStatus:
        try:
            usage = psutil.disk_usage(path)
            free_gb = usage.free / (1024**3)
            total_gb = usage.total / (1024**3)
            percent_used = usage.percent

            healthy = percent_used < 95
            message = f"{free_gb:.1f}GB free of {total_gb:.1f}GB ({percent_used}% used)"

            if percent_used > 85:
                message += " [WARNING: Low disk space]"

            return HealthStatus(
                name="disk",
                healthy=healthy,
                message=message,
                details={
                    "path": path,
                    "free_gb": round(free_gb, 2),
                    "total_gb": round(total_gb, 2),
                    "percent_used": percent_used,
                },
            )
        except Exception as e:
            return HealthStatus(
                name="disk",
                healthy=False,
                message=f"Cannot check disk: {e}",
            )

    def _check_environment(self) -> HealthStatus:
        missing: list[str] = []
        present: list[str] = []

        env_vars = ["GITHUB_TOKEN", "ANTHROPIC_API_KEY", "CONVEX_URL"]

        for var in env_vars:
            if os.environ.get(var):
                present.append(var)
            else:
                missing.append(var)

        healthy = True
        message = f"{len(present)} configured, {len(missing)} optional"

        return HealthStatus(
            name="environment",
            healthy=healthy,
            message=message,
            details={
                "present": present,
                "missing": missing,
            },
        )

    def _check_filesystem(self) -> HealthStatus:
        issues: list[str] = []

        home = Path.home()
        if not home.exists():
            issues.append(f"Home directory not accessible: {home}")

        cache_dir = home / ".scholardevclaw" / "cache"
        if not cache_dir.exists():
            try:
                cache_dir.mkdir(parents=True, exist_ok=True)
            except Exception as e:
                issues.append(f"Cannot create cache directory: {e}")

        temp_dir = Path("/tmp")
        if not temp_dir.exists():
            issues.append("Temp directory not accessible")

        healthy = len(issues) == 0
        message = "Filesystem accessible" if healthy else "; ".join(issues)

        return HealthStatus(
            name="filesystem",
            healthy=healthy,
            message=message,
            details={"issues": issues},
        )


class LivenessProbe:
    """Simple liveness probe for health endpoints."""

    def __init__(self, timeout_seconds: float = 30.0):
        self.timeout_seconds = timeout_seconds
        self.last_heartbeat = time.time()

    def heartbeat(self) -> None:
        self.last_heartbeat = time.time()

    def is_alive(self) -> bool:
        return time.time() - self.last_heartbeat < self.timeout_seconds

    def check(self) -> dict[str, Any]:
        return {
            "alive": self.is_alive(),
            "last_heartbeat": datetime.fromtimestamp(self.last_heartbeat).isoformat(),
            "seconds_since_heartbeat": time.time() - self.last_heartbeat,
        }


class ReadinessProbe:
    """Readiness probe to check if service can handle requests."""

    def __init__(self):
        self._ready = True
        self._reasons: list[str] = []

    def set_ready(self, ready: bool, reason: str | None = None) -> None:
        self._ready = ready
        if reason and not ready:
            self._reasons.append(reason)
        elif ready:
            self._reasons.clear()

    def is_ready(self) -> bool:
        return self._ready

    def check(self) -> dict[str, Any]:
        return {
            "ready": self._ready,
            "reasons": self._reasons,
        }


health_checker = HealthChecker()
liveness_probe = LivenessProbe()
readiness_probe = ReadinessProbe()
